### PR TITLE
DBZ-4353 Add test for online ddl schema migration

### PR DESCRIPTION
**Pull request by sonne5**
Addressed https://issues.redhat.com/browse/DBZ-4353 to add test cases for online ddl schema migration

A few noticeable changes are:
- bump Vitess test container version to 12.0.0 to match the java sdk version. also online ddl is supported after 10.0.0
- move some sharable setup from `VitessReplicationConnectionIT.java` to `TestHelper.java`
- adding padding `\0` on schema type checking test, this is because Vitess added null padding on fixed length binary column to match what mysql returns https://github.com/vitessio/vitess/pull/8137
- add integration test for schema migration with online ddl strategy: https://github.com/debezium/debezium-connector-vitess/pull/52/files#diff-768a08f8a4779aff486ebdde6d730f5fa4a95913881c8d2b535ff332c301c237R41

### CODE REVIEW

**Comment by sonne5** [comment]
I'm getting weird errors complaining `d[] != d[]`:
```
VitessConnectorIT.shouldReceiveChangesForInsertsWithDifferentDataTypes:119->assertInsert:590->assertInsert:617->AbstractVitessConnectorTest.assertRecordSchemaAndValues:377->AbstractVitessConnectorTest.lambda$assertRecordSchemaAndValues$0:378 Values don't match for field 'binary_col' expected:<d[]> but was:<d[]>
```

@keweishang do you have some idea what might be the issue here? thanks!

**Comment by sonne5** [comment]
I figured it out, the failing test is caused by the padding for binary column introduced at Vitess https://github.com/vitessio/vitess/pull/8137

ptal @shichao-an @jeffchao @gunnarmorling

**Comment by jeffchao** [comment]
@sonne5 can you add some more context in the PR description? I'm wondering about the behavior we didn't catch before accounting for the Vitess padding change.

**Comment by sonne5** [comment]
> @sonne5 can you add some more context in the PR description? I'm wondering about the behavior we didn't catch before accounting for the Vitess padding change.

@jeffchao Sure thing. I added some more context in the description. There is no logic change at all, just adding a test case for schema migration. All other code changes are made either for refactoring to reuse or adjust test case for new Vitess's behavior. lmk if it is clear to you.

**Comment by jeffchao** [comment]
Nice, lgtm.

**Review by jeffchao — APPROVED** [review]
lgtm, but also my first review in this org, so let's hear from @gunnarmorling (when he's online) to see if he has any thoughts as well.

File: src/test/java/io/debezium/connector/vitess/VitessSchemaMigrationIT.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+                    .atMost(Duration.ofSeconds(TestHelper.waitTimeForRecords()))
+                    .until(started::get);
+            consumedMessages.clear();
+            // OnlineDDL is async, hence we issue the command and wait until it finishes
+            String ddlId = TestHelper.applySchema("ALTER TABLE t1 ADD COLUMN name varchar(64)", TestHelper.TEST_UNSHARDED_KEYSPACE);
+            Awaitility
+                    .await()
+                    .atMost(Duration.ofSeconds(TestHelper.waitTimeForRecords()))
+                    .until(() -> {
+                        Thread.sleep(1000);
```
Comment (gunnarmorling) [inline src/test/java/io/debezium/connector/vitess/VitessSchemaMigrationIT.java:79]: Rather use `pollInterval()` / `pollDelay()` from Awaitility?

File: src/test/java/io/debezium/connector/vitess/VtctldConnection.java [hunk 2]
```diff
@@ -97,6 +98,25 @@ public void applyVSchema(String vschema, String keyspace) {
...
     }
 
+    protected String applySchema(String sql, String strategy, String keyspace) {
+        List<String> args = Arrays.asList("ApplySchema", "-ddl_strategy=" + strategy, "-sql=" + sql, keyspace);
+        List<String> results = execVtctl(args, vtctldHost, vtctldPort);
+        LOGGER.info("Schema {} is applied. Result: {}", sql, results);
+        return results.get(0).trim();
+    }
+
+    protected boolean checkOnlineDDLCompleted(String keyspace, String id) {
```
Comment (gunnarmorling) [inline src/test/java/io/debezium/connector/vitess/VtctldConnection.java:108]:
```suggestion
    protected boolean checkOnlineDdlCompleted(String keyspace, String id) {
```

File: src/test/java/io/debezium/connector/vitess/VitessSchemaMigrationIT.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                    "UPDATE",
+                    () -> {
+                        try {
+                            return consumedMessages.poll(SECONDS.toMillis(5), TimeUnit.MILLISECONDS);
+                        }
+                        catch (InterruptedException e) {
+                            return null;
+                        }
+                    });
+            ReplicationMessage dml = message.getMessage();
```
Comment (gunnarmorling) [inline src/test/java/io/debezium/connector/vitess/VitessSchemaMigrationIT.java:95]: So I'm a bit confused about this test; I would rather have expected a new test method added to `VitessConnectorIT`, which asserts the actual `SourceRecord` emitted by the connector, including the schema of the new field. I fear with the currently proposed test we're not making sure that the functionality actually works as intended.

**Review by gunnarmorling — COMMENTED** [review]
Thanks a lot @sonne5; some comments inline. Mostly cosmetics, but I'm not 100% about the testing approach. Is it actually the case that no changes to the connector code itself are needed for this, i.e. were schema changes supported before, and it's solely a matter of adding this test? Do you have a pointer to where in the connector actually handles that?

File: src/test/docker/Dockerfile [hunk 4]
```diff
@@ -1,7 +1,7 @@
 # Use a temporary layer for the build stage.
-FROM vitess/base:v9.0.0 AS base
+FROM vitess/base:v12.0.0 AS base
```
[the remark this answers is not in the document]
Comment (sonne5) [reply src/test/docker/Dockerfile:2]: That's a good call, didn't know we have a CI config to test multiple version!

Vitess has 3-4 major release a year, and looks like debezium is doing 4 minor release and 1 major release a year - so there is a small discrepancy. 

That being said, i think it still makes sense to have CI configured to test multiple version. I cut https://issues.redhat.com/browse/DBZ-4363 to follow it up

(and here is the PR to update the tested version: https://github.com/debezium/debezium.github.io/pull/759)

File: src/test/java/io/debezium/connector/vitess/TestHelper.java [hunk 5]
```diff
@@ -128,6 +140,24 @@ protected static void applyVSchema(String vschemaFile) throws Exception {
         }
     }
 
+    protected static String applySchema(String ddl, String keyspace) throws Exception {
```
[the remark this answers is not in the document]
Comment (sonne5) [reply src/test/java/io/debezium/connector/vitess/TestHelper.java:143]: Yep, removed

File: src/test/java/io/debezium/connector/vitess/TestHelper.java [hunk 6]
```diff
@@ -282,6 +312,73 @@ public static int defaultNumOfColumns() {
         return defaultColumnValues().stream().map(x -> x.getJavaValue()).collect(Collectors.toList());
     }
 
+    protected static List<MessageAndVgtid> awaitMessages(
```
[the remark this answers is not in the document]
Comment (sonne5) [reply src/test/java/io/debezium/connector/vitess/TestHelper.java:315]: Yep, good idea. Updated

File: src/test/java/io/debezium/connector/vitess/VitessSchemaMigrationIT.java (hunk 3 above)
Comment (sonne5) [reply src/test/java/io/debezium/connector/vitess/VitessSchemaMigrationIT.java:95]: Good point, I added another test case in `VitessConnectorIT` and also move this one into `VitessReplicationConnectionIT` - lmk if this makes sense to you :-)

**Comment by sonne5** [comment]
> Is it actually the case that no changes to the connector code itself are needed for this, i.e. were schema changes supported before, and it's solely a matter of adding this test? Do you have a pointer to where in the connector actually handles that?

@gunnarmorling Yes, it was supported before ([code pointer](https://github.com/debezium/debezium-connector-vitess/blob/f5ee71c1dfc9709107aab181ea152813b6af33ec/src/main/java/io/debezium/connector/vitess/connection/VStreamOutputMessageDecoder.java#L325-L354))

When there is a DDL, Vitess will streamed out a `FIELD_EVENT`, this is the signal for the VStream consumer that they should cache the updated schema. In Debezium case, it is the signal for us to update the schema for the table.

As for test cases, there are multiple approach for DDL in Vitess: we can do "ALTER TABLE" via pure sql protocol or do it through "online ddl". The sql protocol approach works except it is unreliable, given Vitess is a sharded cluster: if the DDL failed on a certain shards, the cluster will be in a weird state and hard to track. Online ddl does is similar to gh-ost but it's using Vitess's own replication logic to copy the temp table, control the cutover, etc - thus in theory it has less production impact comparing to gh-ost or pt-osc.

We have ddl test case for the [sql protocol](https://github.com/debezium/debezium-connector-vitess/blob/8fcef955ab32877dd34d23a5ceace03e873ae387/src/test/java/io/debezium/connector/vitess/VitessConnectorIT.java#L146) but we didn't have test coverage for online ddl (which is actually the one most ppl are using)

Let me know if this makes sense to you!

File: src/test/docker/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (gunnarmorling) [reply src/test/docker/Dockerfile:2]: > I cut https://issues.redhat.com/browse/DBZ-4363 to follow it up

Thx!

**Review by gunnarmorling — APPROVED** [review]
LGTM, applying. Thx, @sonne5 et al.!

File: src/test/docker/Dockerfile (hunk 4 above)
[the remark this answers is not in the document]
Comment (sonne5) [reply src/test/docker/Dockerfile:2]: @gunnarmorling one more naive question: taking postgres setup as an example https://github.com/debezium/debezium/blob/main/.github/workflows/postgres-workflow.yml

Could you point me to the place where we are running 6 version of it in CI build? I got a bit lost while trying to figure out the setup for the job matrix here.. 😄

[the remark this answers is not in the document]
Comment (gunnarmorling) [reply src/test/docker/Dockerfile:2]: @sonne5, [replied](https://issues.redhat.com/browse/DBZ-4363?focusedCommentId=19385299&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19385299) on the Jira issue, as the discussion on the merged PR is easy to miss. Let's continue over there if you got more questions. Thx!
